src/wallet/wallet.cpp

```diff
@@ -512,20 +512,49 @@ void CWallet::SetBestChain(const CBlockLocator& loc)
     SetBestChainINTERNAL(walletdb, loc);
 }
 
-std::set<std::pair<libzcash::PaymentAddress, uint256>> CWallet::GetNullifiersForAddresses(const std::set<libzcash::PaymentAddress> & addresses)
+std::set<std::pair<libzcash::PaymentAddress, uint256>> CWallet::GetNullifiersForAddresses(
+        const std::set<libzcash::PaymentAddress> & addresses)
 {
     std::set<std::pair<libzcash::PaymentAddress, uint256>> nullifierSet;
+    // Sapling ivk -> list of addrs map
+    std::map<libzcash::SaplingIncomingViewingKey, std::vector<libzcash::SaplingPaymentAddress>> ivkMap;
+    for (const auto & addr : addresses) {
+        auto saplingAddr = boost::get<libzcash::SaplingPaymentAddress>(&addr);
+        if (saplingAddr != nullptr) {
+            libzcash::SaplingIncomingViewingKey ivk;
+            this->GetSaplingIncomingViewingKey(*saplingAddr, ivk);
+            ivkMap[ivk].push_back(*saplingAddr);
+        }
+    }
     for (const auto & txPair : mapWallet) {
+        // Sprout
         for (const auto & noteDataPair : txPair.second.mapSproutNoteData) {
-            if (noteDataPair.second.nullifier && addresses.count(noteDataPair.second.address)) {
-                nullifierSet.insert(std::make_pair(noteDataPair.second.address, noteDataPair.second.nullifier.get()));
+            auto & noteData = noteDataPair.second;
+            auto & nullifier = noteData.nullifier;
+            auto & address = noteData.address;
+            if (nullifier && addresses.count(address)) {
+                nullifierSet.insert(std::make_pair(address, nullifier.get()));
+            }
+        }
+        // Sapling
+        for (const auto & noteDataPair : txPair.second.mapSaplingNoteData) {
+            auto & noteData = noteDataPair.second;
+            auto & nullifier = noteData.nullifier;
+            auto & ivk = noteData.ivk;
+            if (nullifier && ivkMap.count(ivk)) {
+                for (const auto & addr : ivkMap[ivk]) {
+                    nullifierSet.insert(std::make_pair(addr, nullifier.get()));
+                }
```

Comment:
This also means that a user of `z_listreceivedbyaddress` can't tell which diversified address a payment was received on, despite that being an important design goal of diversified addresses.

Reply:
@daira So @LarryRuane and myself have gone over this. For `z_listreceivedbyaddress` there is no impact because you can only pass in one address, and `GetFilteredNotes` will only return notes for any given address(es) it is passed.  However, there might be an issue with `z_listunspent` because a user can pass in a list of addresses, where these addresses may all be derived from the same spending key but with a different diversifier.  In that situation, `GetNullifiersForAddresses` would create an `ivkMap` where `ivkMap[ivk]` is a vector of the addresses passed in, meaning that the nullifier set returns extra entries which are not correct e.g. `(addr1, nullifier1), (addr2, nullifier1), (addr3, nullifier1)`. However, in this case, `z_listunspent` only uses the nullifier set to set the change flag, and it will look for `(addr1, nullifier1)`. 

Given the above, we don't think we need to make any changes right now for this, and we don't support diversified addresses yet. Does the above satisfy your review comment?  We could open an issue about `GetNullifiersForAddresses` and have review for v2.0.2?

Reply:
You're right; for `z_listreceivedbyaddress` I had missed that the loop at [line 3342](https://github.com/zcash/zcash/pull/3499/files#diff-df7d84ff2f53fcb2a0dc15a3a51e55ceR3342) is over the `saplingEntries` found by `GetFilteredNotes`, not `GetNullifiersForAddresses`.

I don't understand the argument for `z_listunspent` but that may just be because I'm tired right now.

Reply:
In any case, please add a comment to `GetNullifiersForAddresses` explaining the current behaviour and why it is ok for now.

Reply:
@daira, I'm sorry, I'm not seeing what comment to make in `GetNullifiersForAddresses()`. Let me try to go through this with an example. Suppose `GetNullifiersForAddresses()` is called with 4 addresses, `za1`, `za2`, `za7`, `za8`. Let's say four addresses, `za1`-`za4` are diversified from the same spending key, so they have the same ivk, which has value `ivk_1234`. So the first loop will create:
```
    ivkMap[ivk_1234]: (za1, za2)
    ivkMap[ivk_7]:    (za7)
    ivkMap[ivk_8]:    (za8)
```
The third (last) loop iterates all the notes in the wallet, each of which has a nullifier and an ivk:
```
    if (nullifier && ivkMap.count(ivk)) {
        for (const auto & addr : ivkMap[ivk]) {
            nullifierSet.insert(std::make_pair(addr, nullifier.get()));
        }
    }
```
After this loops runs, `nullifierSet` will contain four entries, `za1`, `za2`, `za7`, `za8`, the first two of which will have the same nullifier. If and when this loop encounters notes for `za3` and `za4`, they will not be added to `nullifierSet`, even though they have the same nullifier as `za1` and `za2`, because `ivkMap[]` doesn't contain `za3` and `za4`. Am I missing something?

Reply:
In any case, since this is merged, let's just track this as part of #3547.